lib/usdUfe/ufe/UsdClipboardCommands.cpp

```diff
@@ -0,0 +1,496 @@
+    }
+
+    auto appendToVector = [](const auto&             itemsToAppend,
+                             std::vector<Ufe::Path>& vector,
+                             bool                    useMetadataNames = false) {
+        for (const auto& item : itemsToAppend) {
+            auto itemPath = item->path();
+            if (useMetadataNames) {
+                // If we have the original name, use it.
+                const auto newName
+                    = item->getGroupMetadata(
+                              std::string(kClipboardMetadata), std::string(kNodeName))
+                          .template get<std::string>();
+                if (!newName.empty() && newName != item->nodeName()) {
+                    itemPath = itemPath.sibling(newName);
+                }
+            }
+
+            vector.push_back(itemPath);
+        }
+    };
+
+    for (const auto& dstParentItem : _dstParentItems) {
+        Ufe::PasteClipboardCommand::PasteInfo pasteInfo;
+        pasteInfo.pasteTarget = dstParentItem->path();
+
+        if (!clipboardMaterials.empty()) {
+            if (isMaterialsScope(dstParentItem)) {
+                auto duplicateCmd
+                    = UsdUndoDuplicateSelectionCommand::create(clipboardMaterials, dstParentItem);
+                duplicateCmd->execute();
+                appendToVector(duplicateCmd->targetItems(), pasteInfo.successfulPastes);
+                auto tmpTargetItems = duplicateCmd->targetItems();
+                _targetItems.insert(
+                    _targetItems.end(), tmpTargetItems.begin(), tmpTargetItems.end());
+            } else {
+                appendToVector(clipboardMaterials, pasteInfo.failedPastes, true);
+            }
+        }
+
+        if (!clipboardShaders.empty()) {
+            // If the destination target is a Scope and we have shaders to paste, then before we
+            // have to create a material using their clipboard metadata and use it as paste target.
+            if (isMaterialsScope(dstParentItem)) {
+                auto duplicatedItems = pasteItemsToNewMaterial(dstParentItem, clipboardShaders);
+                appendToVector(duplicatedItems, pasteInfo.successfulPastes);
+                _targetItems.insert(
+                    _targetItems.end(), duplicatedItems.begin(), duplicatedItems.end());
+            } else if (PXR_NS::UsdShadeNodeGraph(dstParentItem->prim())) {
+                auto duplicateCmd
+                    = UsdUndoDuplicateSelectionCommand::create(clipboardShaders, dstParentItem);
+                duplicateCmd->execute();
+                appendToVector(duplicateCmd->targetItems(), pasteInfo.successfulPastes);
+                auto tmpTargetItems = duplicateCmd->targetItems();
+                _targetItems.insert(
+                    _targetItems.end(), tmpTargetItems.begin(), tmpTargetItems.end());
+            } else {
+                appendToVector(clipboardShaders, pasteInfo.failedPastes, true);
```

Comment:
IDK why we could not do the same as for scope for other target prim types? In MayaUSD, we sometimes put material under the prim that uses it, for example. OTOH, I guess this is all existing LookdevX code with their own choices of constraints... IDK if these constraints would apply in 3dsMax, for example.

Reply:
Yes the way that shader/materials are handled all comes from the LookdevX design and original implementation. It is unchanged.